controller/gateway/controller.go

```diff
@@ -655,6 +749,184 @@ func (r *Reconciler) provisionControlPlane(
 	return controlPlane
 }
 
+func (r *Reconciler) provisionKonnectGatewayControlPlane(
+	ctx context.Context,
+	logger logr.Logger,
+	gateway *gwtypes.Gateway,
+	gatewayConfig *GatewayConfiguration,
+) *konnectv1alpha2.KonnectGatewayControlPlane {
+	logger = logger.WithName("konnectGatewayControlPlaneProvisioning")
+
+	log.Trace(logger, "Looking for associated KonnectGatewayControlPlanes")
+	konnectControlPlanes, err := gatewayutils.ListKonnectGatewayControlPlanesForGateway(ctx, r.Client, gateway)
+	if err != nil {
+		log.Debug(logger, fmt.Sprintf("failed listing associated konnect gateway controlplanes - error: %v", err))
+		k8sutils.SetCondition(
+			createControlPlaneCondition(metav1.ConditionFalse, kcfgdataplane.UnableToProvisionReason, err.Error(), gateway.Generation),
+			gatewayConditionsAndListenersAware(gateway),
+		)
+		return nil
+	}
+
+	switch count := len(konnectControlPlanes); {
+	case count == 0:
+		_, err := r.createKonnectGatewayControlPlane(ctx, gateway, gatewayConfig)
+		if err != nil {
+			log.Debug(logger, fmt.Sprintf("KonnectGatewayControlPlane creation failed - error: %v", err))
+			k8sutils.SetCondition(
+				createControlPlaneCondition(metav1.ConditionFalse, kcfgdataplane.UnableToProvisionReason, err.Error(), gateway.Generation),
+				gatewayConditionsAndListenersAware(gateway),
+			)
+		} else {
+			log.Debug(logger, "KonnectGatewayControlPlane created")
+			k8sutils.SetCondition(
+				createControlPlaneCondition(metav1.ConditionFalse, kcfgdataplane.ResourceCreatedOrUpdatedReason, kcfgdataplane.ResourceCreatedMessage, gateway.Generation),
+				gatewayConditionsAndListenersAware(gateway),
+			)
+		}
+		return nil
+	case count > 1:
+		err := fmt.Errorf("KonnectGatewayControlPlanes found: %d, expected: 1, requeuing", count)
+		k8sutils.SetCondition(
+			createControlPlaneCondition(metav1.ConditionFalse, kcfgdataplane.UnableToProvisionReason, err.Error(), gateway.Generation),
+			gatewayConditionsAndListenersAware(gateway),
+		)
+		log.Debug(logger, "reducing KonnectGatewayControlPlanes", "count", count)
+		if rErr := k8sreduce.ReduceKonnectGatewayControlPlane(ctx, r.Client, konnectControlPlanes); rErr != nil {
+			log.Error(logger, rErr, "reducing KonnectGatewayControlPlanes failed")
+			return nil
+		}
+		return nil
+	}
+
+	// If we continue, there is only one konnect gateway controlplane.
+	konnectControlPlane := konnectControlPlanes[0].DeepCopy()
+
+	log.Trace(logger, "Ensuring KonnectGatewayControlPlane config is up to date")
```

Comment:
nit: do we log anyway also if we don't actually check the config yet? 🤔 
Btw, your call, feel free to just dismiss this comment.
```suggestion
	// log.Trace(logger, "Ensuring KonnectGatewayControlPlane config is up to date")
```